catalog/CatalogDatabaseCache.cpp

```diff
@@ -61,15 +61,17 @@ void CatalogDatabaseCache::update(const serialization::CatalogDatabase &proto) {
 
   SpinSharedMutexExclusiveLock<false> lock(relations_mutex_);
   for (int i = 0; i < proto.relations_size(); ++i) {
-    auto relation_schema = make_unique<const CatalogRelationSchema>(proto.relations(i));
-    const relation_id rel_id = relation_schema->getID();
+    const serialization::CatalogRelationSchema &proto_relation = proto.relations(i);
+    const relation_id rel_id = proto_relation.relation_id();
```

Comment:
I didn't actually think of concurrent queries for the refactoring, but thanks for pointing to that. There are three reasons:

1. You don't need mutex protection to read proto. Therefore line 64 and 65 can be moved out of any protected scope. 
2. Finding whether a rel_id is in rel_map_ only requires shared lock. Therefore line 67 - 71 can be moved to a shared lock scope. 
3. Finally when you want to modify rel_map_, you should grab an exclusive lock. 

I think it is fine to delay the refactoring until we have concurrent query support, but you can add a TODO for that.